File: apps/mobile/android/app/src/main/java/com/flagshipserver/app/core/NfcPairReader.kt
// C3 Wave 2 (Android) — testable seam over the NFC reader.
//
// Two records on the tag:
//
//   Record 1: NDEF MIME-typed JSON
//     TNF  = TNF_MIME_MEDIA
//     type = "application/flagship.pair+json"
//     payload = UTF-8 JSON object with the fields:
//       { v: number,
//         stkPub, eBoxPub, nonce, sessionId,           // lowercase hex
//         hint: { mdnsName, cloudRendezvousId, suffix6 } }
//
//   Field-name parity with the iOS sibling
//   (apps/mobile/ios/Sources/FlagshipUI/NFC/NfcPairReader.swift) —
//   bytes-on-the-wire are hex strings; the JSON key names DO NOT carry
//   a `Hex` suffix so a future protocol bump can swap the representation
//   without breaking the key contract.
//
//   Record 2: NDEF MIME-typed raw Ed25519 signature
//     TNF  = TNF_MIME_MEDIA
//     type = "application/flagship.pair.sig"
//     payload = 64 raw bytes — Ed25519_sign(stkPriv, canonicalPair(PAIR))
//
// The wire format intentionally avoids NDEF Well-Known TNFs so a stray
// Android Beam / browser intent never tries to "open" the tag — only an
// app that explicitly hunts for these MIME types reads it. The iOS
// sibling (built in parallel) MUST match this exact shape; any deviation
// gets documented in a code comment + a fixture entry.
//
// LiveNfcPairReader uses enableReaderMode with the standard NFC-A flag
// set; on tag discovery it pulls the NDEF message, parses both records,
// and verifies the signature via NfcPair.verifyPair before resuming the
// caller's coroutine.
//
// MockNfcPairReader is the test seam — it returns a fixture without
// touching any hardware so view-model tests stay deterministic + run
// under Robolectric without a NFC stub.

package com.flagshipserver.app.core

import android.app.Activity
import android.nfc.NdefMessage
import android.nfc.NdefRecord
import android.nfc.NfcAdapter
import android.nfc.Tag
import android.nfc.tech.Ndef
import androidx.activity.ComponentActivity
import kotlinx.coroutines.suspendCancellableCoroutine
import org.json.JSONObject
import kotlin.coroutines.resume

// ── Wire-format constants ──────────────────────────────────────────────

/** MIME type carried in the JSON-encoded PAIR record (record 1). */
const val NFC_PAIR_PAYLOAD_MIME: String = "application/flagship.pair+json"

/** MIME type carried in the raw-bytes Ed25519 signature record (record 2). */
const val NFC_PAIR_SIGNATURE_MIME: String = "application/flagship.pair.sig"

private const val ED25519_SIG_LEN: Int = 64

// ── Result + errors ────────────────────────────────────────────────────

/** Outcome of a successful NFC pair read: the parsed payload + the raw
 *  Ed25519 signature that has ALREADY been verified by the reader. */
data class ReadPairResult(val payload: PairPayload, val signature: ByteArray) {
    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other !is ReadPairResult) return false
        return payload == other.payload && signature.contentEquals(other.signature)
    }

    override fun hashCode(): Int = payload.hashCode() * 31 + signature.contentHashCode()
}

/** Typed failure modes — the view model maps these to user-facing copy. */
sealed interface NfcPairReaderError {
    /** No NFC adapter on the device, or it's disabled in system settings. */
    data object NfcUnavailable : NfcPairReaderError

    /** Caller cancelled (e.g. backed out of the read screen). */
    data object UserCanceled : NfcPairReaderError

    /** Tag dispatched but is not the Flagship-pair shape (no MIME match,
     *  not NDEF, wrong record count, etc.). */
    data object TagFormatUnrecognized : NfcPairReaderError

    /** PAIR record present but its body doesn't parse into a PairPayload. */
    data class MalformedPayload(val reason: String) : NfcPairReaderError

    /** PAIR + SIG records present but verifyPair returned false. */
    data object SignatureMismatch : NfcPairReaderError

    /** Reader timed out without a tap. */
    data object Timeout : NfcPairReaderError
}

/** Carrier exception used to surface reader errors through Result.failure. */
class NfcPairReaderException(val error: NfcPairReaderError, message: String) :
    RuntimeException(message)

// ── Reader interface ───────────────────────────────────────────────────

interface NfcPairReader {
    /**
     * Begin a single-shot NFC tag read using foreground reader-mode
     * dispatch. The activity is needed to wire up the adapter callback.
     *
     * Returns Result.success(parsed) on a tag whose signature checks
     * out, Result.failure(NfcPairReaderException) for every other
     * outcome (no adapter / cancel / malformed / verify fail).
     */
    suspend fun readPair(activity: ComponentActivity): Result<ReadPairResult>
}

// ── JSON ⇄ PairPayload helpers ─────────────────────────────────────────

/** Encode a PairPayload as the canonical JSON record-1 body. Public so
 *  the daemon-side fixtures / Mock reader / cross-language vectors can
 *  build records without re-encoding the fields by hand. */
fun encodePairPayloadJson(payload: PairPayload): String {
    val hint = JSONObject().apply {
        put("mdnsName", payload.hint.mdnsName)
        put("cloudRendezvousId", payload.hint.cloudRendezvousId)
        put("suffix6", payload.hint.suffix6)
    }
    return JSONObject().apply {
        put("v", payload.v)
        put("stkPub", NfcPairHex.encode(payload.stkPub))
        put("eBoxPub", NfcPairHex.encode(payload.eBoxPub))
        put("nonce", NfcPairHex.encode(payload.nonce))
        put("sessionId", NfcPairHex.encode(payload.sessionId))
        put("hint", hint)
    }.toString()
}

/** Decode the JSON record-1 body into a PairPayload. Throws
 *  IllegalArgumentException on any shape mismatch — callers turn that
 *  into NfcPairReaderError.MalformedPayload. */
fun decodePairPayloadJson(json: String): PairPayload {
    val root = try {
        JSONObject(json)
    } catch (t: Throwable) {
        throw IllegalArgumentException("PAIR record is not valid JSON: ${t.message}")
    }
    val v = root.optInt("v", PAIR_PROTOCOL_VERSION)
    val stkPubHex = root.optString("stkPub", "")
    val eBoxPubHex = root.optString("eBoxPub", "")
    val nonceHex = root.optString("nonce", "")
    val sessionIdHex = root.optString("sessionId", "")
    val hintObj = root.optJSONObject("hint")
        ?: throw IllegalArgumentException("PAIR.hint missing")
    val mdnsName = hintObj.optString("mdnsName", "")
    val cloudRendezvousId = hintObj.optString("cloudRendezvousId", "")
    val suffix6 = hintObj.optString("suffix6", "")
    if (stkPubHex.isEmpty() || eBoxPubHex.isEmpty() || nonceHex.isEmpty() ||
        sessionIdHex.isEmpty() || mdnsName.isEmpty() || cloudRendezvousId.isEmpty() ||
        suffix6.isEmpty()
    ) {
        throw IllegalArgumentException("PAIR record missing required field(s)")
    }
    val stkPub = try { NfcPairHex.decode(stkPubHex) } catch (t: Throwable) {
        throw IllegalArgumentException("stkPub is not valid hex")
    }
    val eBoxPub = try { NfcPairHex.decode(eBoxPubHex) } catch (t: Throwable) {
        throw IllegalArgumentException("eBoxPub is not valid hex")
    }
    val nonce = try { NfcPairHex.decode(nonceHex) } catch (t: Throwable) {
        throw IllegalArgumentException("nonce is not valid hex")
    }
    val sessionId = try { NfcPairHex.decode(sessionIdHex) } catch (t: Throwable) {
        throw IllegalArgumentException("sessionId is not valid hex")
    }
    return PairPayload(
        v = v,
        stkPub = stkPub,
        eBoxPub = eBoxPub,
        nonce = nonce,
        sessionId = sessionId,
        hint = PairHint(mdnsName, cloudRendezvousId, suffix6),
    )
}

/**
 * Build an NDEF message that carries a PAIR + SIG record pair. Used by
 * the Mock reader to fabricate a tag fixture; also useful for any
 * future device-to-device emulation experiments + the cross-language
 * conformance suite.
 */
fun buildPairNdefMessage(payload: PairPayload, signature: ByteArray): NdefMessage {
    require(signature.size == ED25519_SIG_LEN) { "SIG must be 64 bytes" }
    val pairRecord = NdefRecord.createMime(
        NFC_PAIR_PAYLOAD_MIME,
        encodePairPayloadJson(payload).toByteArray(Charsets.UTF_8),
    )
    val sigRecord = NdefRecord.createMime(NFC_PAIR_SIGNATURE_MIME, signature)
    return NdefMessage(arrayOf(pairRecord, sigRecord))
}

/**
 * Parse + verify a discovered NDEF message. Returns the verified result
 * on success; throws NfcPairReaderException on any unmet precondition.
 *
 * Exposed (not private) so unit tests can pin the parse logic in
 * isolation from the LiveNfcPairReader's reader-mode coroutine plumbing.
 */
fun parseAndVerifyPairNdef(message: NdefMessage): ReadPairResult {
    val records = message.records ?: emptyArray()
    if (records.size < 2) {
        throw NfcPairReaderException(
            NfcPairReaderError.TagFormatUnrecognized,
            "tag has ${records.size} record(s); need ≥2",
        )
    }
    var pairBytes: ByteArray? = null
    var sigBytes: ByteArray? = null
    for (rec in records) {
        if (rec.tnf != NdefRecord.TNF_MIME_MEDIA) continue
        val mime = String(rec.type, Charsets.UTF_8)
        when (mime) {
            NFC_PAIR_PAYLOAD_MIME -> if (pairBytes == null) pairBytes = rec.payload
            NFC_PAIR_SIGNATURE_MIME -> if (sigBytes == null) sigBytes = rec.payload
        }
    }
    val pb = pairBytes ?: throw NfcPairReaderException(
        NfcPairReaderError.TagFormatUnrecognized,
        "no $NFC_PAIR_PAYLOAD_MIME record",
    )
    val sb = sigBytes ?: throw NfcPairReaderException(
        NfcPairReaderError.TagFormatUnrecognized,
        "no $NFC_PAIR_SIGNATURE_MIME record",
    )
    if (sb.size != ED25519_SIG_LEN) {
        throw NfcPairReaderException(
            NfcPairReaderError.MalformedPayload("SIG must be 64 bytes; got ${sb.size}"),
            "signature wrong length",
        )
    }
    val payload = try {
        decodePairPayloadJson(String(pb, Charsets.UTF_8))
    } catch (t: IllegalArgumentException) {
        throw NfcPairReaderException(
            NfcPairReaderError.MalformedPayload(t.message ?: "malformed PAIR JSON"),
            t.message ?: "malformed PAIR JSON",
        )
    }
    if (!verifyPair(payload, sb)) {
        throw NfcPairReaderException(
            NfcPairReaderError.SignatureMismatch,
            "Ed25519 verifyPair returned false",
        )
    }
    return ReadPairResult(payload, sb)
}

// ── Live impl (Android reader-mode) ────────────────────────────────────

class LiveNfcPairReader : NfcPairReader {
    override suspend fun readPair(activity: ComponentActivity): Result<ReadPairResult> {
        val adapter = NfcAdapter.getDefaultAdapter(activity)
        if (adapter == null || !adapter.isEnabled) {
            return Result.failure(
                NfcPairReaderException(
                    NfcPairReaderError.NfcUnavailable,
                    "no NFC adapter (or disabled in system settings)",
                ),
            )
        }
        return suspendCancellableCoroutine { cont ->
            var resumed = false
            val flags = NfcAdapter.FLAG_READER_NFC_A or
                NfcAdapter.FLAG_READER_NFC_B or
                NfcAdapter.FLAG_READER_NFC_F or
                NfcAdapter.FLAG_READER_NFC_V
            val callback = NfcAdapter.ReaderCallback { tag: Tag ->
                if (resumed) return@ReaderCallback
                val outcome = try {
                    val ndef = Ndef.get(tag) ?: throw NfcPairReaderException(
                        NfcPairReaderError.TagFormatUnrecognized,
                        "tag does not present NDEF tech",
                    )
                    ndef.connect()
                    val message = try {
                        ndef.ndefMessage ?: throw NfcPairReaderException(
                            NfcPairReaderError.TagFormatUnrecognized,
                            "tag has no NDEF message",
                        )
                    } finally {
                        try { ndef.close() } catch (_: Throwable) { /* best-effort */ }
                    }
                    Result.success(parseAndVerifyPairNdef(message))
                } catch (e: NfcPairReaderException) {
                    Result.failure(e)
                } catch (t: Throwable) {
                    Result.failure(
                        NfcPairReaderException(
                            NfcPairReaderError.TagFormatUnrecognized,
                            t.message ?: "tag read failed",
                        ),
                    )
                }
                resumed = true
                try {
                    adapter.disableReaderMode(activity)
                } catch (_: Throwable) { /* best-effort */ }
                cont.resume(outcome)
            }
            try {
                adapter.enableReaderMode(activity as Activity, callback, flags, null)
            } catch (t: Throwable) {
                cont.resume(
                    Result.failure(
                        NfcPairReaderException(
                            NfcPairReaderError.NfcUnavailable,
                            "enableReaderMode threw: ${t.message}",
                        ),
                    ),
                )
                return@suspendCancellableCoroutine
            }
            cont.invokeOnCancellation {
                if (!resumed) {
                    resumed = true
                    try { adapter.disableReaderMode(activity) } catch (_: Throwable) { /* best-effort */ }
                }
            }
        }
    }
}

// ── Mock impl (tests / dev-mode) ───────────────────────────────────────

/**
 * Returns a pre-baked Result on every readPair call without touching
 * any Android NFC API. The activity argument is ignored — view-model
 * tests pass an empty stand-in.
 */
class MockNfcPairReader(
    private val outcome: Result<ReadPairResult>,
) : NfcPairReader {
    /** Number of times readPair has been called. Tests use this to
     *  confirm the VM only kicks off a read on startTap. */
    @Volatile var callCount: Int = 0
        private set

    override suspend fun readPair(activity: ComponentActivity): Result<ReadPairResult> {
        callCount += 1
        return outcome
    }
}

File: apps/mobile/ios/Sources/FlagshipUI/NFC/NfcPairReader.swift
import Foundation
#if canImport(CoreNFC)
import CoreNFC
#endif
import FlagshipCore

/// C3 — phone-side NFC reader for the retail-box pairing tap.
///
/// The branded box exposes an NDEF tag (writable in production hardware,
/// read-only fallback shipped to dev kits) containing TWO records:
///
///   1. **PAIR** — `application/flagship.pair+json`, the JSON-encoded
///      PairPayload (`v`, `stkPub`, `eBoxPub`, `nonce`, `sessionId`,
///      `hint{mdnsName,cloudRendezvousId,suffix6}`).
///   2. **SIG**  — `application/flagship.pair.sig`, raw 64-byte Ed25519
///      signature over `canonicalPair(payload)`.
///
/// **Wire format assumption**: PAIR-as-JSON-MIME and SIG-as-raw-MIME is
/// the simplest scheme that round-trips through CoreNFC's NDEF surface
/// without inventing a new TNF. The box-side daemon (N-BOX track) is not
/// yet shipping a final emitter; if it lands with a different scheme
/// (e.g. PAIR-as-bytes via canonicalPair instead of JSON, or both
/// records concatenated in a single MIME blob), the parser below is the
/// single place to adjust. Bytes-on-wire stay protocol-stable because
/// `verifyPair` runs against canonical bytes that we re-derive locally
/// from the parsed PairPayload — the JSON envelope is only a transport.
///
/// The reader is single-shot: it opens an NFCTagReaderSession, polls
/// ISO14443 (covers MIFARE / NTAG / DESFire — the cheapest hardware
/// options for a sticker tag), reads the first detected tag's NDEF
/// message, parses both records, calls `verifyPair`, then invalidates
/// the session. The continuation resolves with `ReadPairResult` or
/// throws `NfcPairReaderError`.
///
/// Surfaces consumed by `NfcPairViewModel`; the live impl is gated
/// behind `#if canImport(CoreNFC)` so test targets and the macOS
/// SPM build (which has no CoreNFC) still compile.

public struct ReadPairResult: Sendable, Equatable {
    public let payload: PairPayload
    /// Raw 64-byte Ed25519 SIG record bytes.
    public let signature: Data

    public init(payload: PairPayload, signature: Data) {
        self.payload = payload
        self.signature = signature
    }
}

public enum NfcPairReaderError: Error, Equatable, Sendable {
    /// Core NFC isn't available (no NFC chip, or running on a platform
    /// that doesn't import CoreNFC — e.g. macOS / Simulator).
    case sessionUnavailable
    /// User dismissed the system NFC sheet before a tag was read.
    case userCanceled
    /// Tag responded but it wasn't NDEF (or NDEF was empty).
    case tagFormatUnrecognized
    /// NDEF must contain exactly PAIR + SIG (= 2 records). Anything else
    /// is either an unrelated tag or a malformed box-side emitter.
    case multipleRecords(Int)
    /// PAIR record present but JSON didn't deserialize into a valid
    /// PairPayload (missing field, wrong hex length, etc.).
    case malformedPayload(String)
    /// `verifyPair` returned false — either tampered, or the wrong tag.
    case signatureMismatch
    /// Session invalidated by Core NFC for a timeout reason.
    case timeout
}

public protocol NfcPairReaderProtocol: Sendable {
    /// Begin a single-shot NFC tag read. Resolves with the first tag's
    /// PAIR + SIG content on success, throws `NfcPairReaderError`.
    func readPair() async throws -> ReadPairResult
}

// ────────────────────────────────────────────────────────────────────────
// MIME types — single source of truth for the box-side emitter to match.

public enum NfcPairWireFormat {
    public static let pairMimeType = "application/flagship.pair+json"
    public static let sigMimeType = "application/flagship.pair.sig"
}

// ────────────────────────────────────────────────────────────────────────
// Parser — pure function over an NDEF message's (type, payload) records.
// Pulled out of the live reader so MockNfcPairReader callers + tests can
// exercise the same wire format without instantiating Core NFC.

public enum NfcPairWireParser {
    /// Parse a pre-extracted (type, payload) list into a `ReadPairResult`.
    /// Each entry is `(mimeType, payloadBytes)` — exactly the shape
    /// `NFCNDEFPayload.type` + `payload` give us, but without the
    /// CoreNFC import so tests stay platform-portable.
    public static func parse(records: [(type: String, payload: Data)]) throws -> ReadPairResult {
        if records.count != 2 {
            throw NfcPairReaderError.multipleRecords(records.count)
        }
        let pairRec = records.first { $0.type == NfcPairWireFormat.pairMimeType }
        let sigRec = records.first { $0.type == NfcPairWireFormat.sigMimeType }
        guard let pairRec else {
            throw NfcPairReaderError.malformedPayload("missing PAIR record (\(NfcPairWireFormat.pairMimeType))")
        }
        guard let sigRec else {
            throw NfcPairReaderError.malformedPayload("missing SIG record (\(NfcPairWireFormat.sigMimeType))")
        }
        if sigRec.payload.count != 64 {
            throw NfcPairReaderError.malformedPayload("SIG must be 64 bytes (Ed25519); got \(sigRec.payload.count)")
        }
        let payload = try decodePairJSON(pairRec.payload)
        if !verifyPair(payload, signature: sigRec.payload) {
            throw NfcPairReaderError.signatureMismatch
        }
        return ReadPairResult(payload: payload, signature: sigRec.payload)
    }

    /// JSON shape:
    /// `{ v, stkPub:hex, eBoxPub:hex, nonce:hex, sessionId:hex,
    ///    hint:{ mdnsName, cloudRendezvousId, suffix6 } }`.
    static func decodePairJSON(_ data: Data) throws -> PairPayload {
        guard let obj = try? JSONSerialization.jsonObject(with: data) as? [String: Any] else {
            throw NfcPairReaderError.malformedPayload("PAIR record is not JSON")
        }
        guard let stkPubHex = obj["stkPub"] as? String,
              let eBoxPubHex = obj["eBoxPub"] as? String,
              let nonceHex = obj["nonce"] as? String,
              let sessionIdHex = obj["sessionId"] as? String,
              let hint = obj["hint"] as? [String: Any],
              let mdnsName = hint["mdnsName"] as? String,
              let cloudRendezvousId = hint["cloudRendezvousId"] as? String,
              let suffix6 = hint["suffix6"] as? String
        else {
            throw NfcPairReaderError.malformedPayload("PAIR JSON missing required field")
        }
        let v: Int
        if let n = obj["v"] as? Int { v = n }
        else if let n = obj["v"] as? Double { v = Int(n) }
        else { v = PAIR_PROTOCOL_VERSION }

        let stkPub = NfcPairHex.decode(stkPubHex)
        let eBoxPub = NfcPairHex.decode(eBoxPubHex)
        let nonce = NfcPairHex.decode(nonceHex)
        let sessionId = NfcPairHex.decode(sessionIdHex)
        if stkPub.count != 32 || eBoxPub.count != 32 {
            throw NfcPairReaderError.malformedPayload("stkPub/eBoxPub must be 32 bytes")
        }
        if nonce.count != 16 || sessionId.count != 16 {
            throw NfcPairReaderError.malformedPayload("nonce/sessionId must be 16 bytes")
        }
        return PairPayload(
            v: v,
            stkPub: stkPub,
            eBoxPub: eBoxPub,
            nonce: nonce,
            sessionId: sessionId,
            hint: PairHint(mdnsName: mdnsName, cloudRendezvousId: cloudRendezvousId, suffix6: suffix6)
        )
    }

    /// Inverse of `decodePairJSON` — used by tests + by the dev-mode
    /// "emit a fake tag" helper. Kept here so the canonical wire format
    /// lives in one place.
    public static func encodePairJSON(_ p: PairPayload) throws -> Data {
        let obj: [String: Any] = [
            "v": p.v,
            "stkPub": NfcPairHex.encode(p.stkPub),
            "eBoxPub": NfcPairHex.encode(p.eBoxPub),
            "nonce": NfcPairHex.encode(p.nonce),
            "sessionId": NfcPairHex.encode(p.sessionId),
            "hint": [
                "mdnsName": p.hint.mdnsName,
                "cloudRendezvousId": p.hint.cloudRendezvousId,
                "suffix6": p.hint.suffix6,
            ]
        ]
        // .sortedKeys keeps the test golden bytes stable across Swift
        // runtimes; it has no effect on the parser.
        return try JSONSerialization.data(withJSONObject: obj, options: [.sortedKeys])
    }
}

// ────────────────────────────────────────────────────────────────────────
// Mock — scripted result for tests + dev-mode dry runs.

public final class MockNfcPairReader: NfcPairReaderProtocol, @unchecked Sendable {
    private let scripted: Result<ReadPairResult, NfcPairReaderError>
    public private(set) var callCount: Int = 0

    public init(result: Result<ReadPairResult, NfcPairReaderError>) {
        self.scripted = result
    }

    public func readPair() async throws -> ReadPairResult {
        callCount += 1
        switch scripted {
        case .success(let r): return r
        case .failure(let e): throw e
        }
    }
}

// ────────────────────────────────────────────────────────────────────────
// Live — Core NFC backed.

#if canImport(CoreNFC)

public final class LiveNfcPairReader: NSObject, NfcPairReaderProtocol, NFCTagReaderSessionDelegate, @unchecked Sendable {

    private let alertMessage: String
    private var session: NFCTagReaderSession?
    private var continuation: CheckedContinuation<ReadPairResult, Error>?

    public init(alertMessage: String = "Hold your phone to the Flagship box.") {
        self.alertMessage = alertMessage
    }

    public func readPair() async throws -> ReadPairResult {
        guard NFCTagReaderSession.readingAvailable else {
            throw NfcPairReaderError.sessionUnavailable
        }
        return try await withCheckedThrowingContinuation { (cont: CheckedContinuation<ReadPairResult, Error>) in
            self.continuation = cont
            let s = NFCTagReaderSession(
                pollingOption: [.iso14443],
                delegate: self,
                queue: nil
            )
            s?.alertMessage = self.alertMessage
            if let s {
                self.session = s
                s.begin()
            } else {
                cont.resume(throwing: NfcPairReaderError.sessionUnavailable)
                self.continuation = nil
            }
        }
    }

    // MARK: NFCTagReaderSessionDelegate

    public func tagReaderSessionDidBecomeActive(_ session: NFCTagReaderSession) {
        // No-op; the user-facing message is already set via session.alertMessage.
    }

    public func tagReaderSession(_ session: NFCTagReaderSession, didInvalidateWithError error: Error) {
        let mapped = mapInvalidationError(error)
        finish(.failure(mapped))
    }

    public func tagReaderSession(_ session: NFCTagReaderSession, didDetect tags: [NFCTag]) {
        guard let first = tags.first else {
            session.invalidate(errorMessage: "No tag detected.")
            return
        }
        session.connect(to: first) { [weak self] err in
            guard let self else { return }
            if let err {
                session.invalidate(errorMessage: "Couldn't connect: \(err.localizedDescription)")
                return
            }
            let ndefTag = self.asNdefTag(first)
            guard let ndefTag else {
                session.invalidate(errorMessage: "Tag isn't NDEF.")
                self.finish(.failure(.tagFormatUnrecognized))
                return
            }
            ndefTag.queryNDEFStatus { status, _, queryErr in
                if let queryErr {
                    session.invalidate(errorMessage: "NDEF query failed: \(queryErr.localizedDescription)")
                    return
                }
                if status == .notSupported {
                    session.invalidate(errorMessage: "Tag isn't NDEF.")
                    self.finish(.failure(.tagFormatUnrecognized))
                    return
                }
                ndefTag.readNDEF { message, readErr in
                    if let readErr {
                        session.invalidate(errorMessage: "Couldn't read tag: \(readErr.localizedDescription)")
                        return
                    }
                    guard let message else {
                        session.invalidate(errorMessage: "Tag is empty.")
                        self.finish(.failure(.tagFormatUnrecognized))
                        return
                    }
                    do {
                        let records = message.records.map { rec in
                            (type: String(data: rec.type, encoding: .utf8) ?? "", payload: rec.payload)
                        }
                        let result = try NfcPairWireParser.parse(records: records)
                        session.alertMessage = "Box paired."
                        session.invalidate()
                        self.finish(.success(result))
                    } catch let e as NfcPairReaderError {
                        session.invalidate(errorMessage: "Pair failed.")
                        self.finish(.failure(e))
                    } catch {
                        session.invalidate(errorMessage: "Pair failed.")
                        self.finish(.failure(.malformedPayload(error.localizedDescription)))
                    }
                }
            }
        }
    }

    // MARK: helpers

    private func asNdefTag(_ tag: NFCTag) -> NFCNDEFTag? {
        switch tag {
        case .miFare(let t): return t
        case .iso7816(let t): return t
        case .iso15693(let t): return t
        case .feliCa(let t): return t
        @unknown default: return nil
        }
    }

    private func mapInvalidationError(_ error: Error) -> NfcPairReaderError {
        let nsErr = error as NSError
        if nsErr.domain == NFCReaderError.errorDomain {
            switch nsErr.code {
            case NFCReaderError.readerSessionInvalidationErrorUserCanceled.rawValue:
                return .userCanceled
            case NFCReaderError.readerSessionInvalidationErrorSessionTimeout.rawValue:
                return .timeout
            default:
                return .tagFormatUnrecognized
            }
        }
        return .tagFormatUnrecognized
    }

    private func finish(_ result: Result<ReadPairResult, NfcPairReaderError>) {
        guard let cont = continuation else { return }
        continuation = nil
        session = nil
        switch result {
        case .success(let r): cont.resume(returning: r)
        case .failure(let e): cont.resume(throwing: e)
        }
    }
}

#else

/// Platform fallback (macOS SPM build, Simulator) so call sites compile.
/// `readPair` always throws `.sessionUnavailable`.
public final class LiveNfcPairReader: NfcPairReaderProtocol, @unchecked Sendable {
    public init(alertMessage: String = "") {}
    public func readPair() async throws -> ReadPairResult {
        throw NfcPairReaderError.sessionUnavailable
    }
}

#endif
